File: sdk/typescript/src/internal/dependencies.ts
import {
  canonicalizeJsonValue,
  parseJson,
  type JsonObject,
  type JsonValue,
} from "./jsoncanon"

export const DEPENDENCY_INDEX_FORMAT_VERSION = 0 as const
export const DEPENDENCY_MATERIALIZER_VERSION = "helmr.dependencies.v0" as const

const maxDependencyIndexSizeBytes = 4096
const maxPackageGraphSizeBytes = 16777216
const maxPackageManagerVersionBytes = 64
const sha256DigestPattern = /^sha256:[0-9a-f]{64}$/
const packageManagerVersionPattern = /^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-(0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*)(\.(0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*))*)?$/
const textEncoder = new TextEncoder()

export type DependencyArchitecture = "aarch64" | "x86_64"
export type PackageManagerName = "bun" | "npm"

export interface DependencyPackageManager {
  readonly name: PackageManagerName
  readonly version: string
}

export interface DependencyLockfile {
  readonly name: "bun.lock" | "package-lock.json"
  readonly digest: string
}

export interface DependencyIndex {
  readonly formatVersion: 0
  readonly packageManager: DependencyPackageManager
  readonly lockfile: DependencyLockfile
  readonly localManifestsDigest: string
  readonly packageGraphDigest: string
  readonly packageGraphSizeBytes: number
  readonly materializerVersion: typeof DEPENDENCY_MATERIALIZER_VERSION
  readonly runtimeDigest: string
  readonly architecture: DependencyArchitecture
}

export function parseDependencyIndex(raw: string | Uint8Array): DependencyIndex {
  const input = typeof raw === "string" ? textEncoder.encode(raw) : raw
  if (input.length === 0 || input.length > maxDependencyIndexSizeBytes) {
    throw new Error(`dependency index size must be between 1 and ${maxDependencyIndexSizeBytes} bytes`)
  }
  const value = parseJson(raw)
  const canonical = canonicalizeJsonValue(value)
  if (!bytesEqual(input, canonical)) {
    throw new Error("dependency index is not RFC 8785 canonical JSON")
  }
  return validateDependencyIndexValue(value)
}

export function canonicalDependencyIndex(index: DependencyIndex): Uint8Array {
  validateDependencyIndex(index)
  const canonical = canonicalizeJsonValue(index as unknown as JsonValue)
  if (canonical.length > maxDependencyIndexSizeBytes) {
    throw new Error(`dependency index size must be at most ${maxDependencyIndexSizeBytes} bytes`)
  }
  return canonical
}

export function validateDependencyIndex(index: DependencyIndex): void {
  canonicalizeJsonValue(index as unknown as JsonValue)
  validateDependencyIndexValue(index as unknown as JsonValue)
}

function validateDependencyIndexValue(value: JsonValue): DependencyIndex {
  const root = requireObject(value, "dependency index")
  requireKeys(
    root,
    [
      "architecture",
      "formatVersion",
      "localManifestsDigest",
      "lockfile",
      "materializerVersion",
      "packageGraphDigest",
      "packageGraphSizeBytes",
      "packageManager",
      "runtimeDigest",
    ],
    "dependency index",
  )
  if (root["formatVersion"] !== DEPENDENCY_INDEX_FORMAT_VERSION) {
    throw new Error(`dependency index formatVersion must be ${DEPENDENCY_INDEX_FORMAT_VERSION}`)
  }
  const managerValue = requireObject(root["packageManager"], "dependency index packageManager")
  requireKeys(managerValue, ["name", "version"], "dependency index packageManager")
  const managerName = requirePackageManagerName(managerValue["name"])
  const managerVersion = requireString(managerValue["version"], "dependency index packageManager.version")
  const versionMatch = packageManagerVersionPattern.exec(managerVersion)
  if (
    textEncoder.encode(managerVersion).length > maxPackageManagerVersionBytes ||
    versionMatch?.[0] !== managerVersion
  ) {
    throw new Error("dependency index packageManager.version is not an admitted SemVer")
  }

  const lockfileValue = requireObject(root["lockfile"], "dependency index lockfile")
  requireKeys(lockfileValue, ["digest", "name"], "dependency index lockfile")
  const lockfileName = requireString(lockfileValue["name"], "dependency index lockfile.name")
  const expectedLockfileName = managerName === "bun" ? "bun.lock" : "package-lock.json"
  if (lockfileName !== expectedLockfileName) {
    throw new Error(`dependency index lockfile.name must be ${expectedLockfileName}`)
  }
  const packageGraphSizeBytes = requirePositiveSafeInteger(
    root["packageGraphSizeBytes"],
    "dependency index packageGraphSizeBytes",
  )
  if (packageGraphSizeBytes > maxPackageGraphSizeBytes) {
    throw new Error(`dependency index packageGraphSizeBytes must be at most ${maxPackageGraphSizeBytes}`)
  }
  if (root["materializerVersion"] !== DEPENDENCY_MATERIALIZER_VERSION) {
    throw new Error(`dependency index materializerVersion must be ${DEPENDENCY_MATERIALIZER_VERSION}`)
  }
  const architecture = root["architecture"]
  if (architecture !== "aarch64" && architecture !== "x86_64") {
    throw new Error(`dependency index architecture ${JSON.stringify(architecture)} is unsupported`)
  }
  return {
    formatVersion: DEPENDENCY_INDEX_FORMAT_VERSION,
    packageManager: { name: managerName, version: managerVersion },
    lockfile: {
      name: lockfileName,
      digest: requireDigest(lockfileValue["digest"], "dependency index lockfile.digest"),
    },
    localManifestsDigest: requireDigest(
      root["localManifestsDigest"],
      "dependency index localManifestsDigest",
    ),
    packageGraphDigest: requireDigest(root["packageGraphDigest"], "dependency index packageGraphDigest"),
    packageGraphSizeBytes,
    materializerVersion: DEPENDENCY_MATERIALIZER_VERSION,
    runtimeDigest: requireDigest(root["runtimeDigest"], "dependency index runtimeDigest"),
    architecture,
  }
}

function requireObject(value: JsonValue | undefined, label: string): JsonObject {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`${label} must be an object`)
  }
  return value as JsonObject
}

function requireString(value: JsonValue | undefined, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} must be a string`)
  }
  return value
}

function requireDigest(value: JsonValue | undefined, label: string): string {
  const digest = requireString(value, label)
  if (!sha256DigestPattern.test(digest)) {
    throw new Error(`${label} must be a lowercase SHA-256 digest`)
  }
  return digest
}

function requirePackageManagerName(value: JsonValue | undefined): PackageManagerName {
  if (value !== "bun" && value !== "npm") {
    throw new Error(`dependency index packageManager.name ${JSON.stringify(value)} is unsupported`)
  }
  return value
}

function requirePositiveSafeInteger(value: JsonValue | undefined, label: string): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 1) {
    throw new Error(`${label} must be a positive JavaScript-safe integer`)
  }
  return value
}

function requireKeys(value: JsonObject, expected: readonly string[], label: string): void {
  const keys = Object.keys(value).sort()
  if (keys.length !== expected.length || keys.some((key, index) => key !== expected[index])) {
    throw new Error(`${label} has unknown or missing members`)
  }
}

function bytesEqual(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((value, index) => value === right[index])
}

File: internal/deployment/dependencies.go
package deployment

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"

	"github.com/helmrdotdev/helmr/internal/jsoncanon"
)

const (
	DependencyIndexFormatVersion  = 0
	DependencyCacheFormatVersion  = 0
	DependencyMaterializerVersion = "helmr.dependencies.v0"
	PackageManagerBun             = PackageManagerName("bun")
	PackageManagerNPM             = PackageManagerName("npm")
	maxDependencyIndexSizeBytes   = 4096
	maxPackageManagerVersionBytes = 64
	dependencyCacheKeyDomain      = "helmr.deployment-program-dependencies-cache.v0\x00"
)

var packageManagerVersionPattern = regexp.MustCompile(`^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)(-(0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*)(\.(0|[1-9][0-9]*|[0-9]*[A-Za-z-][0-9A-Za-z-]*))*)?$`)

type PackageManagerName string

type PackageManager struct {
	Name    PackageManagerName `json:"name"`
	Version string             `json:"version"`
}

type DependencyLockfile struct {
	Name   string `json:"name"`
	Digest string `json:"digest"`
}

type DependencyIndex struct {
	FormatVersion         int                 `json:"formatVersion"`
	DependencyPlanDigest  string              `json:"dependencyPlanDigest"`
	PackageManager        PackageManager      `json:"packageManager"`
	Lockfile              DependencyLockfile  `json:"lockfile"`
	LocalManifestsDigest  string              `json:"localManifestsDigest"`
	PackageGraphDigest    string              `json:"packageGraphDigest"`
	PackageGraphSizeBytes int64               `json:"packageGraphSizeBytes"`
	MaterializerVersion   string              `json:"materializerVersion"`
	RuntimeDigest         string              `json:"runtimeDigest"`
	Architecture          RuntimeArchitecture `json:"architecture"`
}

type DependencyCacheInput struct {
	FormatVersion        int                 `json:"formatVersion"`
	DependencyPlanDigest string              `json:"dependencyPlanDigest"`
	PackageManager       PackageManager      `json:"packageManager"`
	Lockfile             DependencyLockfile  `json:"lockfile"`
	LocalManifestsDigest string              `json:"localManifestsDigest"`
	MaterializerVersion  string              `json:"materializerVersion"`
	RuntimeDigest        string              `json:"runtimeDigest"`
	Architecture         RuntimeArchitecture `json:"architecture"`
}

func ParseDependencyIndex(raw []byte) (DependencyIndex, error) {
	if len(raw) == 0 || len(raw) > maxDependencyIndexSizeBytes {
		return DependencyIndex{}, fmt.Errorf("dependency index size is outside [1,%d]", maxDependencyIndexSizeBytes)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return DependencyIndex{}, fmt.Errorf("canonicalize dependency index: %w", err)
	}
	if !bytes.Equal(raw, canonical) {
		return DependencyIndex{}, fmt.Errorf("dependency index is not RFC 8785 canonical JSON")
	}

	var index DependencyIndex
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&index); err != nil {
		return DependencyIndex{}, fmt.Errorf("decode dependency index: %w", err)
	}
	if err := ensureEOF(decoder, "dependency index"); err != nil {
		return DependencyIndex{}, err
	}
	if err := ValidateDependencyIndex(index); err != nil {
		return DependencyIndex{}, err
	}
	complete, err := CanonicalDependencyIndex(index)
	if err != nil {
		return DependencyIndex{}, err
	}
	if !bytes.Equal(raw, complete) {
		return DependencyIndex{}, fmt.Errorf("dependency index does not match the complete canonical v0 shape")
	}
	return index, nil
}

func CanonicalDependencyIndex(index DependencyIndex) ([]byte, error) {
	if err := ValidateDependencyIndex(index); err != nil {
		return nil, err
	}
	raw, err := json.Marshal(index)
	if err != nil {
		return nil, fmt.Errorf("encode dependency index: %w", err)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("canonicalize dependency index: %w", err)
	}
	if len(canonical) > maxDependencyIndexSizeBytes {
		return nil, fmt.Errorf("dependency index size is outside [1,%d]", maxDependencyIndexSizeBytes)
	}
	return canonical, nil
}

func ValidateDependencyIndex(index DependencyIndex) error {
	if index.FormatVersion != DependencyIndexFormatVersion {
		return fmt.Errorf("dependency index formatVersion = %d, want %d", index.FormatVersion, DependencyIndexFormatVersion)
	}
	if err := validateDependencyInputs(
		index.DependencyPlanDigest,
		index.PackageManager,
		index.Lockfile,
		index.LocalManifestsDigest,
		index.MaterializerVersion,
		index.RuntimeDigest,
		index.Architecture,
		"dependency index",
	); err != nil {
		return err
	}
	if !sha256DigestPattern.MatchString(index.PackageGraphDigest) {
		return fmt.Errorf("dependency index packageGraphDigest is not a lowercase SHA-256 digest")
	}
	if index.PackageGraphSizeBytes < 1 || index.PackageGraphSizeBytes > maxProgramFileSizeBytes {
		return fmt.Errorf("dependency index packageGraphSizeBytes is outside [1,%d]", maxProgramFileSizeBytes)
	}
	return nil
}

func CanonicalDependencyCacheInput(input DependencyCacheInput) ([]byte, error) {
	if err := ValidateDependencyCacheInput(input); err != nil {
		return nil, err
	}
	raw, err := json.Marshal(input)
	if err != nil {
		return nil, fmt.Errorf("encode dependency cache input: %w", err)
	}
	canonical, err := jsoncanon.Transform(raw)
	if err != nil {
		return nil, fmt.Errorf("canonicalize dependency cache input: %w", err)
	}
	return canonical, nil
}

func DependencyCacheKey(input DependencyCacheInput) (string, error) {
	canonical, err := CanonicalDependencyCacheInput(input)
	if err != nil {
		return "", err
	}
	digest := domainDigest(dependencyCacheKeyDomain, canonical)
	return "sha256:" + hex.EncodeToString(digest[:]), nil
}

func ValidateDependencyCacheInput(input DependencyCacheInput) error {
	if input.FormatVersion != DependencyCacheFormatVersion {
		return fmt.Errorf(
			"dependency cache input formatVersion = %d, want %d",
			input.FormatVersion,
			DependencyCacheFormatVersion,
		)
	}
	return validateDependencyInputs(
		input.DependencyPlanDigest,
		input.PackageManager,
		input.Lockfile,
		input.LocalManifestsDigest,
		input.MaterializerVersion,
		input.RuntimeDigest,
		input.Architecture,
		"dependency cache input",
	)
}

func validateDependencyInputs(
	dependencyPlanDigest string,
	manager PackageManager,
	lockfile DependencyLockfile,
	localManifestsDigest string,
	materializerVersion string,
	runtimeDigest string,
	architecture RuntimeArchitecture,
	label string,
) error {
	if !sha256DigestPattern.MatchString(dependencyPlanDigest) {
		return fmt.Errorf("%s dependencyPlanDigest is not a lowercase SHA-256 digest", label)
	}
	if manager.Name != PackageManagerBun && manager.Name != PackageManagerNPM {
		return fmt.Errorf("%s packageManager.name %q is unsupported", label, manager.Name)
	}
	if len(manager.Version) == 0 ||
		len(manager.Version) > maxPackageManagerVersionBytes ||
		!packageManagerVersionPattern.MatchString(manager.Version) {
		return fmt.Errorf("%s packageManager.version %q is not an admitted SemVer", label, manager.Version)
	}
	wantLockfile := "bun.lock"
	if manager.Name == PackageManagerNPM {
		wantLockfile = "package-lock.json"
	}
	if lockfile.Name != wantLockfile {
		return fmt.Errorf("%s lockfile.name = %q, want %q", label, lockfile.Name, wantLockfile)
	}
	if !sha256DigestPattern.MatchString(lockfile.Digest) {
		return fmt.Errorf("%s lockfile.digest is not a lowercase SHA-256 digest", label)
	}
	if !sha256DigestPattern.MatchString(localManifestsDigest) {
		return fmt.Errorf("%s localManifestsDigest is not a lowercase SHA-256 digest", label)
	}
	if materializerVersion != DependencyMaterializerVersion {
		return fmt.Errorf(
			"%s materializerVersion = %q, want %q",
			label,
			materializerVersion,
			DependencyMaterializerVersion,
		)
	}
	if !sha256DigestPattern.MatchString(runtimeDigest) {
		return fmt.Errorf("%s runtimeDigest is not a lowercase SHA-256 digest", label)
	}
	if !validArchitecture(architecture) {
		return fmt.Errorf("%s architecture %q is unsupported", label, architecture)
	}
	return nil
}
